### PR TITLE
Add missing documentation for static IP options

### PR DESCRIPTION
**Pull request by aboch**
- This is a follow up after merge of #19001 to add the missing documentation.
- And also a chance for reviewers to ask for modifications.

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

**Comment by aboch** [comment]
ping @thaJeztah @moxiegirl

File: docs/reference/commandline/run.md [hunk 1]
```diff
@@ -330,6 +330,13 @@ This adds the `busybox` container to the `mynet` network.
 $ docker run -itd --net=my-multihost-network busybox
 ```
 
+You can also choose the IP addresses for the container with `--ip` and `--ip6`
+flags when you start the container on a user-defined network.
+
+```bash
+$ docker run -itd --net=my-multihost-network --ip=10.10.9.75 busybox
```
Comment (thaJeztah) [inline docs/reference/commandline/run.md]: do we need `-it` here? (I realize its in the other examples as well)

Comment (aboch) [inline docs/reference/commandline/run.md]: Thanks, will take care of that

**Comment by thaJeztah** [comment]
Can you also;

- Update the man page; https://github.com/docker/docker/blob/master/man/docker-run.1.md
- Perhaps add an example for `--ip6`?

I don't think this option can be set on `docker create` and `docker start`?

**Comment by thaJeztah** [comment]
Oh, and the network connect man-page; https://github.com/docker/docker/blob/master/man/docker-network-connect.1.md

File: docs/reference/commandline/run.md (hunk 1 above)
Comment (aboch) [inline docs/reference/commandline/run.md]: Actually if we do not start it with that option set, given no command is specified after the image name, the container will go in stopped state right away and user won't be able to inspect a running container, checking the settings while replicating the documentation examples as tutorial.

**Comment by aboch** [comment]
Ahh... Thanks, I only looked at the docs directory...

**Comment by aboch** [comment]
Thanks @thaJeztah for pointing me to the man pages. I think I now took care of them.
And you are correct about `docker start`. While `docker create` is instead valid, thanks for mentioning it.

File: docs/reference/commandline/run.md (hunk 1 above)
Comment (thaJeztah) [inline docs/reference/commandline/run.md]: ah, makes sense (I usually use `nginx` for testing, but I don't think we use it in our docs)

**Comment by thaJeztah** [comment]
thanks! in that case; the `docker create` online docs also need to be updated; https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md :smile:

File: docs/reference/commandline/run.md (hunk 1 above)
Comment (vdemeester) [inline docs/reference/commandline/run.md]: @aboch I think using the `nginx` (or an image that *stays up* and will not need `-it`, just `-d`) image would make more sense.

Comment (thaJeztah) [inline docs/reference/commandline/run.md]: @vdemeester its quite a bit larger though. Perhaps we need a really tiny go webserver for documentation purposes?

Comment (vdemeester) [inline docs/reference/commandline/run.md]: @thaJeztah yep you are right ; something like this : https://github.com/jwilder/whoami ?

**Comment by vdemeester** [comment]
@aboch `docs/reference/commandline/network-connect.md` should be updated too :stuck_out_tongue:

**Comment by aboch** [comment]
Thanks @thaJeztah, taken care of cmd line refs for `docker cereate`.
@vdemester I already took care of that file in my past commit.

**Comment by vdemeester** [comment]
@aboch I meant this part : https://github.com/docker/docker/pull/19286/files#diff-03988418670020a7af9b0c25af75ec01R48 :stuck_out_tongue_closed_eyes:.

**Comment by aboch** [comment]
@vdemeester Ah, it's a copy of the other one. Let me take care of that. Thanks.

File: docs/reference/commandline/network_connect.md [hunk 2]
```diff
@@ -46,11 +46,22 @@ $ docker network connect --link container1:c1 multi-host-network container2
...
-it. The container's IP address is not guaranteed to remain the same when a
-stopped container rejoins the network, unless you specified one when you run
-`docker network connect` command.
+Paused containers remain connected and can be revealed by a `network inspect`.
+When the container is stopped, it does not appear on the network until you restart
+it. If specified, the container's IP address(es) will be reapplied (if still available)
+when a stopped container rejoins the network. One way to guarantee that the container
+will be assigned the same IP addresses when it rejoins the network after a stop
+or a disconnect, is to specify the `--ip-range` when creating the network and choose
+the static IP address(es) from outside the range. This will assure that the IP address
```
Comment (thaJeztah) [inline docs/reference/commandline/network_connect.md]: s/assure/ensure/ ?

File: docs/reference/commandline/network_connect.md [hunk 3]
```diff
@@ -46,11 +46,22 @@ $ docker network connect --link container1:c1 multi-host-network container2
...
-the container is stopped, it does not appear on the network until you restart
-it. The container's IP address is not guaranteed to remain the same when a
-stopped container rejoins the network, unless you specified one when you run
-`docker network connect` command.
+Paused containers remain connected and can be revealed by a `network inspect`.
+When the container is stopped, it does not appear on the network until you restart
+it. If specified, the container's IP address(es) will be reapplied (if still available)
+when a stopped container rejoins the network. One way to guarantee that the container
+will be assigned the same IP addresses when it rejoins the network after a stop
+or a disconnect, is to specify the `--ip-range` when creating the network and choose
```
Comment (thaJeztah) [inline docs/reference/commandline/network_connect.md]: can you add a comma before "and choose"?

File: man/docker-network-connect.1.md [hunk 4]
```diff
@@ -22,14 +22,26 @@ $ docker network connect multi-host-network container1
...
-Paused containers remain connected and a revealed by a `network inspect`. When
-the container is stopped, it does not appear on the network until you restart
-it. The container's IP address is not guaranteed to remain the same when a
-stopped container rejoins the network.
+Paused containers remain connected and can be revealed by a `network inspect`.
+When the container is stopped, it does not appear on the network until you restart
+it. If specified, the container's IP address(es) will be reapplied (if still available)
+when a stopped container rejoins the network. One way to guarantee that the container
+will be assigned the same IP addresses when it rejoins the network after a stop
+or a disconnect, is to specify the `--ip-range` when creating the network and choose
```
Comment (thaJeztah) [inline man/docker-network-connect.1.md]: same here; comma before "and choose"

File: man/docker-network-connect.1.md [hunk 5]
```diff
@@ -22,14 +22,26 @@ $ docker network connect multi-host-network container1
...
+the static IP address(es) from outside the range. This will assure that the IP address
```
Comment (thaJeztah) [inline man/docker-network-connect.1.md]: s/assure/ensure/

**Comment by thaJeztah** [comment]
Thanks @aboch, some minor nits, but LGTM after those are addressed

**Comment by aboch** [comment]
Thanks @thaJeztah.Taken care of your latest comments.

**Comment by thaJeztah** [comment]
LGTM!

ping @vdemeester PTAL :)

**Comment by vdemeester** [comment]
LGTM 🐮

**Comment by bboreham** [comment]
The wording "if still available" does not make clear what will happen if the address is not available: will it fail to start the container or will it choose a different address?

(reading the libnetwork code, Docker's implementation seems to fail the request if it cannot honor the "preferred IP")

**Comment by aboch** [comment]
@bboreham 
You are correct. If the user configured IP address cannot be given to the container, docker will fail the container start.

**Comment by thaJeztah** [comment]
@bboreham if you have suggestions for better wording; can you open a PR? :+1: 
Note that these docs will also still be refined (see the "docs/revisit" label); we currently don't have the bandwidth for that, so any help is welcome

**Comment by bboreham** [comment]
@thaJeztah see #19721 and #19723

**Pull request merged**